File: txt_to_89t.py
import abc
import dataclasses
import os
import typing
import warnings
from enum import IntEnum, StrEnum

txt_file = "test/test.txt"


class SupportedFormat(StrEnum):
    TI92P = "**TI92P*"
    TI89 = "**TI89**"
    TI92 = "**TI92**"

    def supports_text(self) -> bool:
        # cls = self.__class__
        return True

    def supports_image(self) -> bool:
        cls = self.__class__
        return self in (cls.TI92P, cls.TI89)


class DataType(IntEnum):
    TEXT = 0x0B
    NOTEFOLIO = 0x1C
    IMAGE = 0x10


class StoreType(IntEnum):
    RAM = 0
    RAM_LOCKED = 1
    ARCHIVE = 2


@dataclasses.dataclass
class TI_Writer(abc.ABC):
    format: SupportedFormat
    store_type: StoreType
    folder_name: str
    variable_name: str
    comment: str
    data_type: DataType

    @abc.abstractmethod
    def get_content_file_size(self) -> int:
        # text:
        #   0x12 + textContent.length();
        # image:
        #   return 0x13 + bitmap.length / 8;
        pass

    @abc.abstractmethod
    def get_checksum(self) -> int:
        # text:
        #  return (char) (sum + 0xE0 + 0x1);
        # image:
        #   return (char) (sum);
        pass

    @abc.abstractmethod
    def write_content(self, stream: typing.BinaryIO) -> None:
        pass

    def __post_init__(self) -> None:
        @dataclasses.dataclass
        class VarSize:
            name: str
            size: int
            outer: TI_Writer

            @property
            def value_get(self) -> str:
                return getattr(self.outer, self.name)

            def value_set(self, v: str) -> None:
                setattr(self.outer, self.name, v)

            def pad(self, n: int = 1) -> None:
                self.value_set(self.value_get + "\0")

        var_sizes = (
            VarSize("folder_name", 8, self),
            VarSize("variable_name", 8, self),
            VarSize("comment", 40, self),
        )

        for vs in var_sizes:
            if not vs.value_get or len(vs.value_get) > vs.size:
                raise ValueError(vs.name)
            while len(vs.value_get) < vs.size:
                vs.pad()

    @classmethod
    def write_number(cls, stream: typing.BinaryIO, i: int, bytes: int) -> None:
        if i < 0 or i >= (1 << (8 * bytes)):
            raise ValueError
        stream.write(i.to_bytes(bytes, "big"))

    @classmethod
    def write_byte(cls, stream: typing.BinaryIO, i: int) -> None:
        cls.write_number(stream, i, 1)

    @classmethod
    def write_short(cls, stream: typing.BinaryIO, i: int) -> None:
        cls.write_number(stream, i, 2)

    @classmethod
    def write_int(cls, stream: typing.BinaryIO, i: int) -> None:
        cls.write_number(stream, i, 4)

    @classmethod
    def write_string(cls, stream: typing.BinaryIO, s: str, min_size: int = 0) -> None:
        if min_size:
            while len(s) < min_size:
                s += "\0"
        stream.write(s.encode("ascii"))

    def write_header(self, stream: typing.BinaryIO) -> None:
        cls = self.__class__

        cls.write_string(stream, self.format.value)
        cls.write_short(stream, 0x1 << 8)
        cls.write_string(stream, self.folder_name, 8)
        cls.write_string(stream, self.comment, 40)
        cls.write_short(stream, 0x1 << 8)  # number of vars = 1
        cls.write_int(stream, 0x52 << 24)  # data adress
        cls.write_string(stream, self.variable_name, 8)
        cls.write_byte(stream, self.data_type.value)  # data type
        cls.write_byte(stream, self.store_type.value)  # store type
        cls.write_short(stream, 0xFF << 8)  # ??

        stream.flush()  # maybe not needed
        file_size = stream.tell() + self.get_content_file_size()

        cls.write_byte(stream, file_size & 0xFF)
        cls.write_byte(stream, (file_size >> 8) & 0xFF)
        cls.write_byte(stream, (file_size >> 24) & 0xFF)
        cls.write_byte(stream, (file_size >> 16) & 0xFF)
        cls.write_short(stream, 0xA55A)  # 2 bytes separator
        cls.write_short(stream, 0x0)  # 2 bytes zeroes

    def write_checksum(self, stream: typing.BinaryIO) -> None:
        cls = self.__class__

        crcsum = self.get_checksum()
        cls.write_byte(stream, crcsum & 0xFF)  # hi a low bytes of crc sum
        cls.write_byte(stream, crcsum >> 8)

    def write(self, stream: typing.BinaryIO) -> None:
        self.write_header(stream)
        self.write_content(stream)
        self.write_checksum(stream)


@dataclasses.dataclass
class TI_TextWriter(TI_Writer):
    text_content: str
    sum: int = dataclasses.field(default=0, init=False)
    data_type: DataType = dataclasses.field(default=DataType.TEXT, init=False)

    def __post_init__(self) -> None:
        super().__post_init__()
        if not self.text_content:
            warnings.warn("Text content is empty", UserWarning)

    @staticmethod
    def unicode_to_ti(c: str) -> int:
        # str is inherently unicode in python
        if len(c) != 1:
            raise ValueError

        m: dict[str, int] = {
            # '\u00D0': 18,
            "\u00bb": 22,
            "\u00bc": 28,
            "\u00bd": 29,
            "\u00be": 30,
            "\u00bf": 31,
            "\u20ac": 128,
            "\u0081": 129,
            "\u201a": 130,
            "\u0192": 131,
            "\u201e": 132,
            "\u2026": 133,
            "\u2020": 134,
            "\u2021": 135,
            "\u02c6": 136,
            "\u2030": 137,
            "\u0160": 138,
            "\u2039": 139,
            "\u0152": 140,
            "\u008d": 141,
            "\u017d": 142,
            "\u008f": 143,
            "\u0090": 144,
            "\u2018": 145,
            "\u2019": 146,
            "\u201c": 147,  # patch
            "\u201d": 148,
            "\u2013": 149,
            "\u2014": 150,
            "\u00c6": 151,
            "\u02dc": 152,
            "\u2122": 153,
            "\u0161": 154,
            "\u203a": 155,
            "\u0153": 156,
            "\u009d": 157,
            "\u017e": 158,
            "\u0178": 159,
            "\u00a1": 160,
            "\u00a2": 161,
            "\u00a3": 162,
            "\u00a4": 163,
            "\u00a5": 164,
            "\u00a6": 165,
            "\u007c": 166,
            # '\u00A7': 167,
            "\u00a7": 168,
            # '\u00A8': 169,
            "\u00aa": 170,
            "\u00d7": 171,
            "\u00a9": 172,
            # '\u00AA': 173,
            "\u00a8": 174,
            "\u00ab": 175,
            "\u00ac": 176,
            "\u00ad": 177,  # plus,minus
            "\u00ae": 178,
            "\u00af": 179,
            "\u00b0": 180,
            "\u00b1": 181,
            # '\u00B6': 182,
            "\u00b2": 183,
            "\u00b3": 184,
            "\u00b4": 185,
            "\u00b5": 186,
            "\u00de": 187,
            "\u00b6": 188,
            "\u00b7": 189,
            "\u00b8": 190,
            "\u00b9": 191,
            "\u00c0": 192,
            "\u00c1": 193,
            "\u00c2": 194,
            "\u00c3": 195,
            "\u00c4": 196,
            "\u00c5": 197,
            # "\u00e6": 198,
            "\u00c7": 199,
            "\u00c8": 200,
            "\u00c9": 201,
            "\u00ca": 202,
            "\u00cb": 203,
            # '\u00CD': 204,
            "\u00cc": 205,
            "\u00cd": 206,
            "\u00cf": 207,
            "\u00d0": 18,
            "\u00d1": 209,
            "\u00d2": 210,
            "\u00d3": 211,
            "\u00d4": 212,
            "\u00d5": 213,
            "\u00d6": 214,
            "\u0078": 215,
            "\u00d8": 216,
            "\u00d9": 217,
            "\u00da": 218,
            "\u00db": 219,
            "\u00dc": 220,
            "\u00dd": 221,
            # "\u00fe": 222,
            "\u00df": 223,
            "\u00e0": 224,
            "\u00e1": 225,
            "\u00e2": 226,
            "\u00e3": 227,
            "\u00e4": 228,
            "\u00e5": 229,
            "\u00e6": 230,
            "\u00e7": 231,
            "\u00e8": 232,
            "\u00e9": 233,
            "\u00ea": 234,
            "\u00eb": 235,
            "\u00ec": 236,
            "\u00ed": 237,
            "\u00ee": 238,
            "\u00ef": 239,
            "\u00f0": 240,
            "\u00f1": 241,
            "\u00f2": 242,
            "\u00f3": 243,
            "\u00f4": 244,
            "\u00f5": 245,
            "\u00f6": 246,
            "\u00f7": 247,
            "\u00f8": 248,
            "\u00f9": 249,
            "\u00fa": 250,
            "\u00fb": 251,
            "\u00fc": 252,
            "\u00fd": 253,
            "\u00fe": 254,
            "\u00ff": 255,
        }

        if c not in m:
            raise ValueError(f"Character {c} is not supported")
        return m[c]

    def get_content_file_size(self) -> int:
        return 0x12 + len(self.text_content)

    def get_checksum(self) -> int:
        return self.sum + 0xE0 + 0x1

    def write_content(self, stream: typing.BinaryIO) -> None:
        cls = self.__class__

        text = self.text_content

        length = len(text) + sum(c == "\n" for c in text)
        length4 = length + 0x4 + 2

        cls.write_int(stream, length4)
        cls.write_byte(stream, 0x0)
        cls.write_byte(stream, 0x1)  # start of text
        cls.write_byte(stream, 0x20)  # TODO: idk if this is needed

        self.sum = (length4 & 0xFF) + ((length4 >> 8) & 0xFF)

        for c in text:
            codes = [ord(c)]

            if codes[0] > 127:
                codes[0] = cls.unicode_to_ti(c)
            elif c == "\n":
                codes = [ord("\r"), 0x20]

            for code in codes:
                self.sum += code
                cls.write_byte(stream, code)
        cls.write_short(stream, 0xE0)  # ende


if __name__ == "__main__":
    with open("test/test.txt", "r") as f:
        content = f.read()

    writer = TI_TextWriter(
        format=SupportedFormat.TI92P,
        store_type=StoreType.ARCHIVE,
        folder_name="main",
        variable_name="test_a",
        comment="comment_b",
        text_content=content,
    )

    with open("test/test_a.89t", "wb") as f:
        writer.write(f)

File: src/txt_to_89t.c
#include <stdlib.h>
#include <stdio.h>
#include <errno.h>
#include <string.h>

#include "bad_hashtable.h"
#include "utf8_reader.h"
#include "file_helpers.h"

#define MIN(a, b) (a > b ? b : a)

static BadHashTable_t unicode_ti_table;

static void fill_unicode_ti_table() {
    bad_ht_initialize(&unicode_ti_table);

    bad_ht_insert(&unicode_ti_table, "\u00BB", 22); // »
    bad_ht_insert(&unicode_ti_table, "\u00BC", 28); // ¼
    bad_ht_insert(&unicode_ti_table, "\u00BD", 29); // ½
    bad_ht_insert(&unicode_ti_table, "\u00BE", 30); // ¾
    bad_ht_insert(&unicode_ti_table, "\u00BF", 31); // ¿
    bad_ht_insert(&unicode_ti_table, "\u20AC", 128); // €
    // bad_ht_insert(&unicode_ti_table, "\u0081", 129); //
    bad_ht_insert(&unicode_ti_table, "\u201A", 130); // ‚
    bad_ht_insert(&unicode_ti_table, "\u0192", 131); // ƒ
    bad_ht_insert(&unicode_ti_table, "\u201E", 132); // „
    bad_ht_insert(&unicode_ti_table, "\u2026", 133); // …
    bad_ht_insert(&unicode_ti_table, "\u2020", 134); // †
    bad_ht_insert(&unicode_ti_table, "\u2021", 135); // ‡
    bad_ht_insert(&unicode_ti_table, "\u02C6", 136); // ˆ
    bad_ht_insert(&unicode_ti_table, "\u2030", 137); // ‰
    bad_ht_insert(&unicode_ti_table, "\u0160", 138); // Š
    bad_ht_insert(&unicode_ti_table, "\u2039", 139); // ‹
    bad_ht_insert(&unicode_ti_table, "\u0152", 140); // Œ
    // bad_ht_insert(&unicode_ti_table, "\u008D", 141); //
    bad_ht_insert(&unicode_ti_table, "\u017D", 142); // Ž
    // bad_ht_insert(&unicode_ti_table, "\u008F", 143); //
    // bad_ht_insert(&unicode_ti_table, "\u0090", 144); //
    bad_ht_insert(&unicode_ti_table, "\u2018", 145); // ‘
    bad_ht_insert(&unicode_ti_table, "\u2019", 146); // ’
    bad_ht_insert(&unicode_ti_table, "\u201C", 147); // “
    bad_ht_insert(&unicode_ti_table, "\u201D", 148); // ”
    bad_ht_insert(&unicode_ti_table, "\u2013", 149); // –
    bad_ht_insert(&unicode_ti_table, "\u2014", 150); // —
    bad_ht_insert(&unicode_ti_table, "\u00C6", 151); // Æ
    bad_ht_insert(&unicode_ti_table, "\u02DC", 152); // ˜
    bad_ht_insert(&unicode_ti_table, "\u2122", 153); // ™
    bad_ht_insert(&unicode_ti_table, "\u0161", 154); // š
    bad_ht_insert(&unicode_ti_table, "\u203A", 155); // ›
    bad_ht_insert(&unicode_ti_table, "\u0153", 156); // œ
    // bad_ht_insert(&unicode_ti_table, "\u009D", 157); //
    bad_ht_insert(&unicode_ti_table, "\u017E", 158); // ž
    bad_ht_insert(&unicode_ti_table, "\u0178", 159); // Ÿ
    bad_ht_insert(&unicode_ti_table, "\u00A1", 160); // ¡
    bad_ht_insert(&unicode_ti_table, "\u00A2", 161); // ¢
    bad_ht_insert(&unicode_ti_table, "\u00A3", 162); // £
    bad_ht_insert(&unicode_ti_table, "\u00A4", 163); // ¤
    bad_ht_insert(&unicode_ti_table, "\u00A5", 164); // ¥
    bad_ht_insert(&unicode_ti_table, "\u00A6", 165); // ¦
    // bad_ht_insert(&unicode_ti_table, "\u007C", 166); // |
    bad_ht_insert(&unicode_ti_table, "\u00A7", 168); // §
    bad_ht_insert(&unicode_ti_table, "\u00AA", 170); // ª
    bad_ht_insert(&unicode_ti_table, "\u00D7", 171); // ×
    bad_ht_insert(&unicode_ti_table, "\u00A9", 172); // ©
    bad_ht_insert(&unicode_ti_table, "\u00A8", 174); // ¨
    bad_ht_insert(&unicode_ti_table, "\u00AB", 175); // «
    bad_ht_insert(&unicode_ti_table, "\u00AC", 176); // ¬
    bad_ht_insert(&unicode_ti_table, "\u00AD", 177); // ­
    bad_ht_insert(&unicode_ti_table, "\u00AE", 178); // ®
    bad_ht_insert(&unicode_ti_table, "\u00AF", 179); // ¯
    bad_ht_insert(&unicode_ti_table, "\u00B0", 180); // °
    bad_ht_insert(&unicode_ti_table, "\u00B1", 181); // ±
    bad_ht_insert(&unicode_ti_table, "\u00B2", 183); // ²
    bad_ht_insert(&unicode_ti_table, "\u00B3", 184); // ³
    bad_ht_insert(&unicode_ti_table, "\u00B4", 185); // ´
    bad_ht_insert(&unicode_ti_table, "\u00B5", 186); // µ
    bad_ht_insert(&unicode_ti_table, "\u00DE", 187); // Þ
    bad_ht_insert(&unicode_ti_table, "\u00B6", 188); // ¶
    bad_ht_insert(&unicode_ti_table, "\u00B7", 189); // ·
    bad_ht_insert(&unicode_ti_table, "\u00B8", 190); // ¸
    bad_ht_insert(&unicode_ti_table, "\u00B9", 191); // ¹
    bad_ht_insert(&unicode_ti_table, "\u00C0", 192); // À
    bad_ht_insert(&unicode_ti_table, "\u00C1", 193); // Á
    bad_ht_insert(&unicode_ti_table, "\u00C2", 194); // Â
    bad_ht_insert(&unicode_ti_table, "\u00C3", 195); // Ã
    bad_ht_insert(&unicode_ti_table, "\u00C4", 196); // Ä
    bad_ht_insert(&unicode_ti_table, "\u00C5", 197); // Å
    bad_ht_insert(&unicode_ti_table, "\u00C7", 199); // Ç
    bad_ht_insert(&unicode_ti_table, "\u00C8", 200); // È
    bad_ht_insert(&unicode_ti_table, "\u00C9", 201); // É
    bad_ht_insert(&unicode_ti_table, "\u00CA", 202); // Ê
    bad_ht_insert(&unicode_ti_table, "\u00CB", 203); // Ë
    bad_ht_insert(&unicode_ti_table, "\u00CC", 205); // Ì
    bad_ht_insert(&unicode_ti_table, "\u00CD", 206); // Í
    bad_ht_insert(&unicode_ti_table, "\u00CF", 207); // Ï
    bad_ht_insert(&unicode_ti_table, "\u00D0", 18); // Ð
    bad_ht_insert(&unicode_ti_table, "\u00D1", 209); // Ñ
    bad_ht_insert(&unicode_ti_table, "\u00D2", 210); // Ò
    bad_ht_insert(&unicode_ti_table, "\u00D3", 211); // Ó
    bad_ht_insert(&unicode_ti_table, "\u00D4", 212); // Ô
    bad_ht_insert(&unicode_ti_table, "\u00D5", 213); // Õ
    bad_ht_insert(&unicode_ti_table, "\u00D6", 214); // Ö
    // bad_ht_insert(&unicode_ti_table, "\u0078", 215); // x
    bad_ht_insert(&unicode_ti_table, "\u00D8", 216); // Ø
    bad_ht_insert(&unicode_ti_table, "\u00D9", 217); // Ù
    bad_ht_insert(&unicode_ti_table, "\u00DA", 218); // Ú
    bad_ht_insert(&unicode_ti_table, "\u00DB", 219); // Û
    bad_ht_insert(&unicode_ti_table, "\u00DC", 220); // Ü
    bad_ht_insert(&unicode_ti_table, "\u00DD", 221); // Ý
    bad_ht_insert(&unicode_ti_table, "\u00DF", 223); // ß
    bad_ht_insert(&unicode_ti_table, "\u00E0", 224); // à
    bad_ht_insert(&unicode_ti_table, "\u00E1", 225); // á
    bad_ht_insert(&unicode_ti_table, "\u00E2", 226); // â
    bad_ht_insert(&unicode_ti_table, "\u00E3", 227); // ã
    bad_ht_insert(&unicode_ti_table, "\u00E4", 228); // ä
    bad_ht_insert(&unicode_ti_table, "\u00E5", 229); // å
    bad_ht_insert(&unicode_ti_table, "\u00E6", 230); // æ
    bad_ht_insert(&unicode_ti_table, "\u00E7", 231); // ç
    bad_ht_insert(&unicode_ti_table, "\u00E8", 232); // è
    bad_ht_insert(&unicode_ti_table, "\u00E9", 233); // é
    bad_ht_insert(&unicode_ti_table, "\u00EA", 234); // ê
    bad_ht_insert(&unicode_ti_table, "\u00EB", 235); // ë
    bad_ht_insert(&unicode_ti_table, "\u00EC", 236); // ì
    bad_ht_insert(&unicode_ti_table, "\u00ED", 237); // í
    bad_ht_insert(&unicode_ti_table, "\u00EE", 238); // î
    bad_ht_insert(&unicode_ti_table, "\u00EF", 239); // ï
    bad_ht_insert(&unicode_ti_table, "\u00F0", 240); // ð
    bad_ht_insert(&unicode_ti_table, "\u00F1", 241); // ñ
    bad_ht_insert(&unicode_ti_table, "\u00F2", 242); // ò
    bad_ht_insert(&unicode_ti_table, "\u00F3", 243); // ó
    bad_ht_insert(&unicode_ti_table, "\u00F4", 244); // ô
    bad_ht_insert(&unicode_ti_table, "\u00F5", 245); // õ
    bad_ht_insert(&unicode_ti_table, "\u00F6", 246); // ö
    bad_ht_insert(&unicode_ti_table, "\u00F7", 247); // ÷
    bad_ht_insert(&unicode_ti_table, "\u00F8", 248); // ø
    bad_ht_insert(&unicode_ti_table, "\u00F9", 249); // ù
    bad_ht_insert(&unicode_ti_table, "\u00FA", 250); // ú
    bad_ht_insert(&unicode_ti_table, "\u00FB", 251); // û
    bad_ht_insert(&unicode_ti_table, "\u00FC", 252); // ü
    bad_ht_insert(&unicode_ti_table, "\u00FD", 253); // ý
    bad_ht_insert(&unicode_ti_table, "\u00FE", 254); // þ
    bad_ht_insert(&unicode_ti_table, "\u00FF", 255); // ÿ
}

// SupportedFormat
typedef enum {
    TI92P, TI89, TI92
} SupportedFormat;
static const char *_supported_format_strs[] = {
    "**TI92P*", "**TI89**", "**TI92**"
};
static const char *supported_format_to_str(SupportedFormat sf) {
    switch (sf) {
    case TI92P: return _supported_format_strs[0];
    case TI89: return _supported_format_strs[1];
    case TI92: return _supported_format_strs[2];
    default:
        fprintf(stderr, "unreachable on %s:%d\n",__FILE__, __LINE__);
        exit(1); // unreachable
    }
}

// DataType
typedef enum {
    TEXT = 0x0B,
    NOTEFOLIO = 0x1C,
    IMAGE = 0x10
} DataType;

// StoreType
typedef enum {
    RAM = 0,
    RAM_LOCKED = 1,
    ARCHIVE = 2
} StoreType;

// TIWriterConfig
#define FOLDER_NAME_LENGTH 8
#define VARIABLE_NAME_LENGTH 8
#define COMMENT_LENGTH 40
typedef struct {
    SupportedFormat format;
    StoreType store_type;
    const char *folder_name;
    const char *variable_name;
    const char *comment;
} TIWriterConfig;

static size_t write_ubyte(FILE *fptr, uint8_t i) {
    fwrite(&i, 1, sizeof(uint8_t), fptr);
    check_write_file_error(fptr);
    return sizeof(uint8_t);
}

static size_t write_ushort(FILE *fptr, uint16_t i) {
    // fwrite(&i, 1, sizeof(uint16_t), fptr);
    // check_write_file_error(fptr);
    // return sizeof(uint16_t);
    return write_endian(fptr, &i, sizeof(uint16_t), BIGENDIAN);
}

static size_t write_uint(FILE *fptr, uint32_t i) {
    // fwrite(&i, 1, sizeof(uint32_t), fptr);
    // check_write_file_error(fptr);
    // return sizeof(uint32_t);
    return write_endian(fptr, &i, sizeof(uint32_t), BIGENDIAN);
}

static size_t write_chararr_fixed_size(FILE *fptr, const char *str, size_t size, size_t fixed_size) {
    fwrite(str, MIN(size, fixed_size), sizeof(char), fptr);
    check_write_file_error(fptr);
    for (size_t i = size; i < fixed_size; ++i) {
        fputc(0x0, fptr);
        check_write_file_error(fptr);
    }
    return sizeof(char) * fixed_size;
}

static size_t write_chararr(FILE *fptr, const char *str, size_t size) {
    return write_chararr_fixed_size(fptr, str, size, size);
}

static size_t write_str_fixed_size(FILE *fptr, const char *str, size_t fixed_size) {
    const size_t size = strlen(str);
    return write_chararr_fixed_size(fptr, str, size, fixed_size);
}

static size_t write_str(FILE *fptr, const char *str) {
    return write_chararr(fptr, str, strlen(str));
}

static void write_header(FILE *fptr, const TIWriterConfig *config, size_t content_file_size, DataType data_type) {
    size_t file_size = 0;

    file_size += write_str(fptr, supported_format_to_str(config->format));
    file_size += write_ushort(fptr, 0x1 << 8);
    file_size += write_str_fixed_size(fptr, config->folder_name, FOLDER_NAME_LENGTH);
    file_size += write_str_fixed_size(fptr, config->comment, COMMENT_LENGTH);
    file_size += write_ushort(fptr, 0x1 << 8); // number of vars = 1
    file_size += write_uint(fptr, 0x52 << 24); // data adress
    // file_size += write_uint(fptr, 0x52 << 8); // data adress
    file_size += write_str_fixed_size(fptr, config->variable_name, VARIABLE_NAME_LENGTH);
    file_size += write_ubyte(fptr, data_type);
    file_size += write_ubyte(fptr, config->store_type);
    // file_size += write_ushort(fptr, 0xFF); // ?
    file_size += write_ushort(fptr, 0xFF << 8); // ?

    file_size += content_file_size + 8;

    write_ubyte(fptr, file_size & 0xFF);
    write_ubyte(fptr, (file_size >> 8) & 0xFF);
    write_ubyte(fptr, (file_size >> 24) & 0xFF);
    write_ubyte(fptr, (file_size >> 16) & 0xFF);
    write_ushort(fptr, 0xA55A); // 2 bytes separator
    write_ushort(fptr, 0x0); // 2 bytes separator
}

static void write_checksum(FILE *fptr, size_t checksum) {
    write_ubyte(fptr, checksum & 0xFF); // hi a low bytes of crc sum
    write_ubyte(fptr, checksum >> 8);
}

// TITextWriter
typedef struct {
    TIWriterConfig config;
    const uint32_t *text_content;
    size_t text_content_size;
} TITextWriter;

static void ti_text_write(FILE *fptr, const TITextWriter *t) {
    // get content size
    unsigned int content_size = t->text_content_size;

    // count newlines
    for (size_t i = 0; i < t->text_content_size; ++i) {
        if (t->text_content[i] == 0x0A) { // newline
            ++content_size;
            printf("R\n");
        }
    }
    printf("tc %lu, c: %u\n", t->text_content_size, content_size);

    write_header(fptr, &t->config, content_size + 7, TEXT);

    write_uint(fptr, content_size);
    write_ubyte(fptr, 0x0);
    write_ubyte(fptr, 0x1); // start of text
    write_ubyte(fptr, 0x20); // TODO: idk if this is needed

    unsigned int sum = (content_size & 0xFF) + ((content_size >> 8) & 0xFF);

    for (size_t i = 0; i < t->text_content_size; ++i) {
        uint32_t c = t->text_content[i];
        printf("%u (%c)\n", c, c);
        if (c <= 0x7F) { // if it's an ascii character
            if (c == 0x0A) { // new line
                const char c2 = 0x0D; // carriage return
                write_ubyte(fptr, c2);
                sum += c2;
                c = 0x20; // space
            } else if (c == 0x0D) { // carriage return -> skip
                continue;
            }
        } else {
            uint8_t value;
            bool special_char = bad_ht_get(&unicode_ti_table, c, &value);
            if (special_char) {
                c = value;
            } else {
                fprintf(stderr, "Invalid code %u\n", c);
                exit(1);
            }
        }
        sum += c;
        write_ubyte(fptr, c);
    }
    write_ushort(fptr, 0xE0); // end of file

    size_t checksum = sum + 0xE0 + 0x1;
    write_checksum(fptr, checksum);
}

void print_usage() {

}

// argument 1 -> .txt source
// argument 2 -> .89t output
// argument 3 -> variable
// argument 4 -> folder

#ifdef DEBUG
#define FNAME_TXT "test/test.txt"
#define FNAME_89T "examples/test_a.89t"
#define VARIABLE "test"
#define FOLDER "main"
#endif

int main(int argc, char *argv[]) {
    #ifndef DEBUG
    if (argc == 1) {
        print_usage();
    } else if (argc < 5) {
        printf("Error: bad arguments");
        print_usage();
    }

    const char *fname_txt = argv[1];
    const char *fname_89t = argv[2];
    const char *variable = argv[3];
    const char *folder = argv[4];
    #else
    const char *fname_txt = FNAME_TXT;
    const char *fname_89t = FNAME_89T;
    const char *variable = VARIABLE;
    const char *folder = FOLDER;
    #endif

    const size_t vlen = strlen(variable);
    const size_t flen = strlen(folder);
    if (vlen == 0 || vlen > VARIABLE_NAME_LENGTH) {
        fprintf(stderr, "Variable has to be at least one char long and up to %u chars long\n", VARIABLE_NAME_LENGTH);
        exit(1);
    }
    if (flen == 0 || flen > FOLDER_NAME_LENGTH) {
        fprintf(stderr, "Folder name has to be at least one char long and up to %u chars long\n", FOLDER_NAME_LENGTH);
        exit(1);
    }

    uint32_t *text_content = NULL;
    const char *a = fname_89t;
    size_t text_content_size = read_utf8(fname_txt, &text_content);

    TITextWriter config = {
        .config = {
            .comment = "comment",
            .folder_name = folder,
            .variable_name = variable,
            .format = TI92P,
            .store_type = ARCHIVE,
        },
        .text_content = text_content,
        .text_content_size = text_content_size,
    };

    // delete old output file
    delete_file_if_exists(fname_89t);
    FILE *out_file_ptr = open_file(fname_89t, "wb");

    ti_text_write(out_file_ptr, &config);

    close_file(out_file_ptr);

    free(text_content);

    puts("done");

    return 0;
}
